lib/Sema/TypeCheckInvertible.cpp

```diff
@@ -0,0 +1,149 @@
+  }
+
+  // Walk the enum elements that have associated values.
+  if (auto enumDecl = dyn_cast<EnumDecl>(nominal)) {
+    for (auto caseDecl : enumDecl->getAllCases()) {
+      for (auto element : caseDecl->getElements()) {
+        if (!element->hasAssociatedValues())
+          continue;
+
+        // Check that the associated value type is Sendable.
+        auto elementType = dc->mapTypeIntoContext(
+            element->getArgumentInterfaceType());
+        if ((*this)(element, elementType))
+          return true;
+      }
+    }
+
+    return false;
+  }
+
+  return false;
+}
+
+bool canBeNoncopyable(Type type) {
+  auto &ctx = type->getASTContext();
+  if (auto *nominal = type->getNominalOrBoundGenericNominal())
+    return evaluateOrDefault(ctx.evaluator, IsNoncopyableRequest{nominal},
+                             /*default=*/false);
+
+  if (auto *expansion = type->getAs<PackExpansionType>()) {
+    return canBeNoncopyable(expansion->getPatternType());
+  }
+
+  // if any components of the tuple are move-only, then the tuple is move-only.
+  if (auto *tupl = type->getCanonicalType()->getAs<TupleType>()) {
+    for (auto eltTy : tupl->getElementTypes())
+      if (canBeNoncopyable(eltTy))
+        return true;
+  }
+
+  return false; // otherwise, the conservative assumption is it's copyable.
+}
+
+bool isNoncopyable(Type type, DeclContext *dc) { return !isCopyable(type, dc); }
+
+bool isCopyable(Type type, DeclContext *dc) {
+  assert(dc);
+  auto contextTy = dc->mapTypeIntoContext(type);
+  auto *module = dc->getParentModule();
+  return TypeChecker::conformsToKnownProtocol(contextTy,
+                                              KnownProtocolKind::Copyable,
+                                              module, false);
+}
+
+/// \returns true iff the given nominal meets the requirements of Copyable>
+bool checkCopyableConformance(Evaluator &evaluator,
+                              NominalTypeDecl *nom) {
+  // All classes can store noncopyable values.
+  if (isa<ClassDecl>(nom))
+    return true;
+
+  // Protocols do not directly define any storage.
+  if (isa<ProtocolDecl>(nom))
+    return true;
+
+  if (isa<BuiltinTupleDecl>(nom))
+    llvm_unreachable("unexpected BuiltinTupleDecl");
+
+  if (auto strct = dyn_cast<StructDecl>(nom)) {
+    for (auto stored : strct->getStoredProperties()) {
+//      stored->getInterfaceType()
+//      ImplicitKnownProtocolConformanceRequest
+    }
+  }
+
+}
+
+ProtocolConformance *deriveConformanceForInvertible(Evaluator &evaluator,
+                                                    NominalTypeDecl *nominal,
+                                                    KnownProtocolKind kp) {
+  auto &ctx = nominal->getASTContext();
+  auto *proto = ctx.getProtocol(kp);
+
+  switch (kp) {
+  case KnownProtocolKind::Copyable: {
+    if (evaluateOrDefault(evaluator, IsNoncopyableRequest{nominal}, false))
+      return nullptr; // it's not Copyable.
+
+    // Otherwise, verify Copyable conformance.
```

Comment:
You should just trust the type's inheritance clause here. Check that the stored properties are copyable in TypeCheckDeclPrimary.cpp's DeclChecker visitor

Reply:
I think you're right here in that I should just return the ProtocolConformance. But the checking of that conformance should probably happen in `TypeChecker::checkConformancesInContext` instead of at the top level / MiscDiagnostics.